### ISSUE TITLE
forEachFeatureAtPixel freezes/super slow performance on ff and ie

### ISSUE THREAD

**Issue by CasparWaara** [openlayers/ol3#4232]
I have 2000+ features on the map and I want to display popup on hover to the features. I have used the forEachFeatureAtPixel  and on chrome it works as expected but with firefox and internet explorer it's completely useless. Whole map freezes and stops working.

I have tried with interactions (I think it uses forEachFeatureAtCoordinate and forEachFeatureAtPixel) and with map.on('pointermove'...). Works ok with Chrome but freezes on FF/IE.

Is there any workarounds, hacks or anything that would/could/should work?

**Comment by elemoine** [comment]
Are you able to provide an example someone could look at?

**Comment by elemoine** [comment]
I am going to close this for now. But feel free to reopen if you can come up with an example with evidences that there's a problem in the library.

**Comment by cvenzin** [comment]
I can confirm that, the example is already on the example page: http://openlayers.org/en/v3.10.1/examples/kml-earthquakes.html?q=toolt
performance is only in chrome acceptable

**Comment by ahocevar** [comment]
Performs great for me on Firefox, Edge and IE 11 for me. Even on a low budget slow CPU tablet. Would be good to see a video of this performance issue.

**Comment by marcjansen** [comment]
It feels notably slower on FF than on Chrome for me.

**Comment by tonio** [comment]
![screenshot from 2015-10-16 15 26 28](https://cloud.githubusercontent.com/assets/21686/10542696/5e06e158-741a-11e5-818c-acb120d0008b.png)

Quite slow for me in FF, too.

**Comment by ahocevar** [comment]
This seems to be a serious issue indeed. http://stackoverflow.com/questions/33246093/very-slow-hover-interactions-in-openlayers-3-with-any-browser-except-chrome reports the same.

**Comment by ahocevar** [comment]
@elemoine, you are most familiar with the hit detection code. Any chance you can take a look at this? Thanks.

**Comment by tonio** [comment]
Some background about performance in FF in https://bugzil.la/1020340 (not sure it’s really related to this issue).

**Comment by elemoine** [comment]
FWIW http://openlayers.org/en/v3.10.1/examples/kml-earthquakes.html works pretty well for me on FF Linux.

**Comment by ahocevar** [comment]
For me as well @elemoine - I think this might depend on the hardware or whether 3D acceleration is available.

**Comment by elemoine** [comment]
It is indeed way faster in Chrome than in Firefox. With the kml-earthquakes example I get a solid 60 fps with Chrome when randomly moving my mouse over the map. I get ~ 20 fps with Firefox. But I find the user experience still acceptable on my hardware.

Note: before wrapX the number of features in the kml-earchquakes map was half. So the hit detection probably felt a lot faster before.

**Comment by ahocevar** [comment]
> Note: before wrapX the number of features in the kml-earchquakes map was half. So the hit detection probably felt a lot faster before.

The number of features is the same. Only for rendering, the replay group is replayed multiple times. For hit detection, the hit detection replay group is only replayed once. Just like without wrapX.

**Comment by elemoine** [comment]
> The number of features is the same. Only for rendering, the replay group is replayed multiple times. For hit detection, the hit detection replay group is only replayed once. Just like without wrapX.

Sh^W :) Thanks the note. It makes sense.

**Comment by tohu12** [comment]
On [this line 1924] (https://github.com/openlayers/ol3/blob/0259977fa95da120b38d56eac0b020190e4cfb22/src/ol/render/canvas/canvasreplay.js#L1924) is checking for alpha information from CanvasContext2D. And this function "context.getImageData(0, 0, 1, 1).data" use very very often for every feature from every layer. And this function is more slower in FF/IE then Chrome/Safari !!!.

I don't know how exactly OL3 work with Vector data, because code is complicated. But i believe, that exist way how optimize this :)

Now, I don´t use ol.interaction.Select({condition: ol.events.condition.pointerMove}) for hover, but i use this optimized code where I call forEachFeatureAtPixel() only every 200ms:
```javascript
...
var timeout;
var self = this;

map.on('pointermove', function (evt) {
    if (evt.dragging  || timeout) return;
    timeout = setTimeout(function () {
        timeout = null;
    }, 200);

    var pixel = this.map.getEventPixel(evt.originalEvent);
    var flData = this.map.forEachFeatureAtPixel(
            pixel,
            function (feature, layer) {
                if (layer) {
                    return {
                        'feature': feature,
                        'layer': layer
                    };
                }
            },
            null,
            function (layer) {
                return (layer === self.myLayer);
            }
        );

    // ... check if is over/out

});
...
```
I have 5 vector layers with many icon features and it's better in FF/IE, but i would like use ol.interaction.Select().

What do you think @ahocevar ?
Thanks

**Comment by debgreb** [comment]
I've put together this [jsfiddle illustrating the poor performance of pointermove](https://jsfiddle.net/debgreb/g7ubk2hc/) in FF/IE, which is severe enough that we are considering disabling the pointermove events for Firefox and IE entirely.  

If you open the fiddle in Chrome/Safari all is well, but in FF/IE it is not even really functional.  I've tried the various work arounds suggested above, but nothing seems to help.  I'd really love to see this problem resolved!

As a side note, suggestions for combining the two pointermove events defined in the jsfiddle are much appreciated.  This is my first ol3 project, and I'm definitely still on the learning curve....

**Comment by marksamman** [comment]
I've combined them for you here: https://jsfiddle.net/gk7bjbqp/1/

It's still performing poorly though, I suggest that you debounce the pointermove event in FF/IE until this has been addressed. I have the same issue in a web app where I have around 500 markers.

**Comment by debgreb** [comment]
Thanks for the new fiddle, marksamman!  Very helpful.

It's good to know that as few as 500 markers triggers the problem, since my example, with over 3,000 features, may have given the illusion that the problem isn't encountered frequently.  Hopefully the ol3 team has a fix for us all soon.

**Comment by rwlnd** [comment]
This is actually quite a blocking issue if you want to provide Firefox support. I tried the above suggestion with the timeout just for sports but that didn't give me the desired result. I have been looking for another event that is triggered when the cursor is stopped, but that doesn't seem to exist. Can you give any update on the current status / priority of this issue or can you provide a proven alternative to achieve the hover effect in FF/IE? Thanks

**Comment by ahocevar** [comment]
@rwlnd If your sports exercise was based on the [mentioned](https://github.com/openlayers/ol3/issues/4232#issuecomment-182013738) fiddle, then there is plenty room for performance optimisation: For highlighting, it is way faster to use a Select interaction than to set a different style on the selected feature.

**Comment by rwlnd** [comment]
@ahocevar, just like the OP, I've been experimenting with the 2 officially documented ways to add the hover effect. The exercise I mentioned was the suggestion posted by @tohu12.

The first way included binding the `pointermove `event of the map and then getting all the features at the pixel position using `forEachFeatureAtPixel `and adding the feature to a separate layer. ([as described in the example](http://openlayers.org/en/v3.14.2/examples/image-vector-layer.html))

The second way included creating an `ol.interaction.Select` interaction and adding it to the map, then binding to the `select ` event of the interaction ([as descibed in this example](http://openlayers.org/en/v3.14.1/examples/select-features.html))

Both of them did not include applying style to features directly. I can also tell you that even without doing anything to the hovered feature(s) itself, Firefox seems to be blocking at getting all the features at the given pixel.

I DO have 24.000 features on a layer. When I reduce the number of features to let's say 500, the performance impact seems less and the laggyness is not so hard, but still present.

ALSO, both ways of applying the hover effect work flawless in Chrome (v49.0.2623.87). My Firefox version is 45.0.1

**Comment by ahocevar** [comment]
So what do you suggest @rwlnd? Consider this a Firefox bug?

**Comment by rwlnd** [comment]
@ahocevar, I was highlighting Firefox as that is the version I am currently supporting but I gave my test project a go in IE 11 and Microsoft Edge (25.10586.0.0) and I am encountering the same performance issues. I am willing to investigate this issue further myself but I am running low on time as I am in the middle of a rewrite of a project from openlayers2 to openlayers3.

**Comment by debgreb** [comment]
@ahocevar , I've also seen the performance problems that @rwlnd is describing, and as he says, they happen not just in FireFox, but also in IE11 and Microsoft Edge.  You can compare the snappy performance of this [fiddle](https://jsfiddle.net/gk7bjbqp/1/) in Chrome/Safari to the nearly unusable performance in FF/IE/Edge.  I would love to be able to solve this by changing the way I've coded the hover event in OpenLayers 3, but have not had any success in doing that on my own.  Any and all suggestions are welcome.

**Comment by ahocevar** [comment]
After profiling this issue, I found that feature hit detection itself is fast. What makes the example slow in Firefox, IE and Edge is the rendering of the county polygons, which is done every time a feature style changes. The reason why the situation is not improved when using an `ol.interaction.Select` with an `ol.events.condition.pointerMove` instead of changing the style of the hovered feature directly is because replaying the render instructions, namely the `context.fill()` instruction, and not the creation of the render replay, is expensive.

You basically have two options to work around this, and I have created a fiddle for both:

1. https://jsfiddle.net/gk7bjbqp/8/ uses an `ol.layer.Image` with `ol.source.ImageVector` for the counties layer. The selected feature is on a separate layer and overlaid on top of the original counties layer.
2. https://jsfiddle.net/gk7bjbqp/9/ is the same as your fiddle above, but without a fill style for the counties layer.

It looks like one improvement we could make on the library/renderer side is to [batch `context.fill()` calls](http://jsperf.com/batch-canvas-stroke-fill-calls/3), but we would still have to stroke polygons separately. This would mean that we'd have to duplicate lineTo/moveTo instructions, and have extra cost in all cases except the special one where all polygons are filled with the same style.

**Comment by bjornharrtell** [comment]
@ahocevar, since #5149 was reverted by #5150 does it mean this issue should be reopened?

**Comment by rwlnd** [comment]
Hey you guys. Sorry I haven't been been updating on this issue any more. At the moment, we are working our way around the `forEachFeatureAtPixel` method provided by openlayers by using a custom workaround to find features on a given location.

To give you an idea of what we do. We directly use the **rbush** and **turf** libraries to build a geospatial index of features when we load features on the map (rbush) and then find the feature at the clicked position (turf). We also measure the velocity of the mouse movent so we don't need to check for changes EVERY pixel but only at low velocity.

One thing that still is a bit tricky is when Linestrings or Points are rendered with a different width or radius but I can confirm you that using these low level libraries, we are around 40ms to respond to a click event in both chrome AND most importantly, also firefox!

I hope this can give you some new insights to tackle this problem because we would love to use the native `forEachFeatureAtPixel` provided by openlayers instead of adding a custom error prone solution..

If you want further details please don't hesitate to contact me.

**Comment by ahocevar** [comment]
@rwlnd, does #5196 fix the issue for you? To try it, configure your layer source with `overlaps: false`.

**Comment by tsauerwein** [comment]
> To give you an idea of what we do. We directly use the rbush and turf libraries to build a geospatial index of features when we load features on the map (rbush) and then find the feature at the clicked position (turf).

ol3 is also using rbush internally, each [vector source](http://openlayers.org/en/master/apidoc/ol.source.Vector.html) has a R-Tree. If you don't want to use `forEachFeatureAtPixel` you can directly call [ol.source.Vector#getFeaturesAtCoordinate](http://openlayers.org/en/master/apidoc/ol.source.Vector.html#getFeaturesAtCoordinate) to query this R-Tree.

>  We also measure the velocity of the mouse movent so we don't need to check for changes EVERY pixel but only at low velocity.

That sounds interesting. Does this give better results that simply using debouncing?

**Comment by marcjansen** [comment]
> That sounds interesting ...

Also intersted in this 👍

**Comment by gholol** [comment]
I'm experiencing similiar problems with the performance on newest firefox. I have a custom map in which i'm using pretty large (1620x1620) static images (svg's), and it's pretty slow on zooming in and out. RAM increases up to 1GB, and can even cause the whole browser to crash. The same map works super fluently in chrome...

**Comment by schwerlaut** [comment]
I'd totally appreciate to see this fixed :+1:

**Comment by ahocevar** [comment]
@schwerlaut Have you tried the fix proposed in #5196?

**Comment by schwerlaut** [comment]
@ahocevar I've tried, but didn't notice any difference.

**Comment by ahocevar** [comment]
@schwerlaut Any chance you could share your code as JSFiddle too? So I can use it for performance tests/comparisons?

**Comment by schwerlaut** [comment]
@ahocevar It's part of a tenant project, so I probably can't share it.
I might try to get a reduced version of it next week, but I can't guarantee that at the moment.
Anyway, may I propose to add a dist folder (and it's content obviously) to your branch for easier testing. That way it could be directly references via node-dependency.

**Comment by ahocevar** [comment]
Thanks @schwerlaut. I'm going to make a build of my batch-fill-stroke branch available and provide a npm package url as soon as the tests there are fixed.

**Comment by ahocevar** [comment]
Published a version of my branch from #5196 with a dist folder: git://github.com/ahocevar/ol3.git#batch-fill-stroke-dist.

**Comment by schwerlaut** [comment]
@ahocevar : For having a ready-to-go-example, you might just use the fiddle posted by you earlier. Just change it to use the js and css from your git-hub-folder (which anyway requires the browser to ignore Mime-conflicts) like that:
https://jsfiddle.net/erdkz98n/1/
A thing I noticed after upgrading to win10 last week: Firefox now performs better then under win 8.1, but Edge is as slow as IE was.

**Comment by ahocevar** [comment]
@schwerlaut I have verified with that jsfiddle that my pull request improves things significantly. I was hoping you could do the same on your platforms.

**Comment by ahocevar** [comment]
@schwerlaut I created a new fiddle which only contains the relevant code:

* Release v3.17.1: http://jsfiddle.net/ub2czho7/
* My branch from #5196 with `overlaps: false`: http://jsfiddle.net/ub2czho7/1/

**Comment by schwerlaut** [comment]
Yah, sorry - my sample-fiddle was nonesense somehow since it wasn't even using the overlaps-option.
Anyway, with FF47 on win10 I don't see that much of a difference between 3.17.1 and the release from #5196. Maybe there is a slight difference, but it is hard to even notice and in direct comparsion to opera or chrome it feels somewhat unusable.

**Comment by michaelletzgus** [comment]
Both version are unusable with FF47.0.1@Debian Linux: After some fills due to MouseOver the browser hangs...

**Comment by ahocevar** [comment]
Ok @schwerlaut and @michaelletzgus, thanks for testing. http://jsfiddle.net/ub2czho7/2/ uses `ol.source.ImageVector`, like in your original fiddle. Does that perform better for you @schwerlaut?

**Comment by schwerlaut** [comment]
@ahocevar I'd say it's even slower.

**Comment by ahocevar** [comment]
Thanks for the reports everyone. What surprises me most is that `ol.source.ImageVector` does not perform better. I'll do more profiling, but it might take a while until I get to it. The goal is to find out where all the time is spent/lost. With a plain vector layer, it appeared to be canvas fill instructions. But with an image layer with `ol.source.ImageVector`, this should not be the case.

**Comment by michaelletzgus** [comment]
http://jsfiddle.net/ub2czho7/2/ won't work at all, the browser hangs after reaching the first element with the mouse without any visual (fill) effect.

**Comment by ahocevar** [comment]
It would help if those experiencing freezes with the latest set of fiddles (http://jsfiddle.net/ub2czho7/1/, http://jsfiddle.net/ub2czho7/2/) could post profiling results.

**Comment by ischas** [comment]
Here is a profile from Win 10 (64 bit) with FF47. It  doesn't freeze, but it's super slow. Similar behaviour with FF Dev Ed 49.0a2.

[vectorlayer-ff.txt](https://github.com/openlayers/ol3/files/360923/vectorlayer-ff.txt)

(I had to rename the json to txt)

**Comment by michaelletzgus** [comment]
New insight from profiling: The hang is caused by browser AddOn "Privacy Badger". With disabled PB, all fiddels work without any hang! Maybe there are other AddOns doing script inspection and/or injections causing the performance problem?
I'll try some detailed profiling with a fresh&clean browser later.

**Comment by ischas** [comment]
@michaelletzgus Cannot confirm this. My FF is running w/o addons.

**Comment by michaelletzgus** [comment]
Without addons it's running slow - but it works and is usable. With enabled Privacy Badger the browser freezes until the stop-the-script-dialog pops up (1-2 minutes).

**Comment by michaelletzgus** [comment]
Here are profiles of fiddle 1 and 2 w/o any AddOns and a profile of fiddle 1 with Privacy Badger enabled.
[ub2czho7_1.json.txt](https://github.com/openlayers/ol3/files/360995/ub2czho7_1.json.txt)
[ub2czho7_1_withPB.json.txt](https://github.com/openlayers/ol3/files/360994/ub2czho7_1_withPB.json.txt)
[ub2czho7_2.json.txt](https://github.com/openlayers/ol3/files/360996/ub2czho7_2.json.txt)

**Comment by ahocevar** [comment]
@ischas Which fiddle did you profile?

**Comment by ischas** [comment]
The second one, haven't seen that there are two.

Here is also a profile of fillde 1:

[vectorlayer-ff-1.txt](https://github.com/openlayers/ol3/files/361038/vectorlayer-ff-1.txt)

**Comment by ahocevar** [comment]
Ok, some interesting results:

TL;DR: Both this ticket (speed up hit detection) and #5196 (speed up rendering) are valid, and not much related.

For @michaelletzgus, #5196 brings performance improvements, because most of the time is spent rendering (and not hit-detecting). Same situation as on my machine. For @ischas, #5196 does not bring performance improvements. Most of the time is spent hit-detecting.

This suggests that depending on the graphics hardware used, we have room for improvement for replaying batches (done with #5196), and for speeding up hit detection (this ticket, no solution yet).

Looking at @michaelletzgus's fiddle with that Privacy Badger plugin, most time is spent internally in the Gecko engine and on JIT compilation. Probably both outside OpenLayers, at least the profile shows no connection to OpenLayers code. The recommendation here is to disable such plugins for sites that use OpenLayers.

**Comment by ahocevar** [comment]
Adding a "Pull request accepted" label. Summarizing the findings from this ticket, options for a fix include:
* https://github.com/openlayers/ol3/issues/4232#issuecomment-223039250: Use a spatial index and point in polygon operations instead of canvas hit detection. Problem: how to handle stroke widths and symbol sizes.
* https://github.com/openlayers/ol3/issues/4232#issuecomment-159916771: See if there is a way to avoid `getImageData` calls in hit detection (or if they are faster in a bigger canvas than the 1x1 pixel one we currently use), because that is the most expensive operation.
* [Canvas hit regions](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/addHitRegion). Problem: currently only supported in Chrome and Firefox.

**Comment by waxenegger** [comment]
I'm going to be rather unproductive and just add my 2 cents so as to perhaps add another experience that can potentially aid somebody who has similar issues.

Using a vector layer with selecting, translating and modifying features that are filled, and use a style function, the hit-detection performs miserably. As somebody who has only read parts of the code and built on it I cannot say what exactly causes the performance hit but looking at the profiler I do know that the replay costs me all the time. Unfortunately I cannot make amends with requirements so the style function, filling and text style are pretty much as given as well as a more than average number of features (potentially exceeding 100K). 

The moment I use the rbush things work smoothly and fast again. It's like day and night. I do not only use it for hit detection but also to create a hierarchy of clusters fitting the viewing extent.

Either way, I think a spatial structure like the rtree can be a good solution for people who have a performance bottleneck with their hit-detection.

**Comment by ahocevar** [comment]
@waxenegger What you're saying does not make 100% sense. You are saying your bottleneck is the replay. Are you rendering your selected features with a different style? If so, it should not make a difference whether you use hit detection or point in polygon checks with an RTree, unless your original vector layer is an `ol.layer.Image` layer with an `ol.source.ImageVector`. The reason is that when rendering a feature with a different style, a whole new map image will be rendered, causing a replay of all `ol.layer.Vector` layers.

**Comment by waxenegger** [comment]
@ahocevar well, I'm not a 100% sure what is causing the problem either ;-)

I do put a different style on a selected feature using a style function but I'm not using ol.source.ImageVector. 

What I observed was the replay as part of hit detection when I looked a the function stack of the profiling output.

I have to also mention that the Translate interaction added  quite a bit to it in my case. Now I inherited from ol.interaction.Translate and overrode respective methods to add stuff I needed but, going back to the original version to test whether I had screwed up showed no performance difference.

Sorry to be somewhat cryptic otherwise, I do have some custom code bits there where I might have done things a bit different (or even not exactly right performance wise) so that drawing conclusions may be more difficult. It's just that for me the replacement of the hitdetection with rtree search did really have a tremendous positive impact.

**Comment by ahocevar** [comment]
@waxenegger Alright, you were talking about the hit detection replay, not the rendering replay. For this, I have another hint: if your layer is a polygon layer, configure a small `renderBuffer`. The default is 100 pixels, which will consider many polygons around the hit point. For a polygon layer, this should be set to the stroke width of the polygon outline (so e.g. 1 or 2).

**Comment by cooperq** [comment]
Interesting that privacy badger would be causing an issue here. Does ol3 make a ton of requests? If so it is possible that privacy badger could slow it down....
Has this been tested on the chrome version of privacy badger at all?

**Comment by michaelletzgus** [comment]
No problems with PB and chromium so far.

@ahocevar Disabling PB for a specific site only does not help, the whole AddOn must be deactivated completely.

**Comment by cooperq** [comment]
Okay with that being said we are going to release the web extensions port of privacy badger for firefox soon which I think should solve this issue on our end. It will share a codebase with privacy badger for chrome.

**Comment by BellisJ** [comment]
Has anyone had any more luck with this issue? This has been a major blocker for implementing hover on a project of mine. It works fine on Mac (OS X El Capitan) and Chrome, but slightly slower on Firefox (though still highly usable). The issue lies in Windows, more specifically our tests have been on Windows 10 while running Firefox; which basically freezes and crashes when attempting to hover with map.on(‘pointermove’, function(e){}) and either a ol.interaction.Select() or map.forEachFeatureAtPixel().

We've tried using both ol.interaction.Select() as well as a map.forEachFeatureAtPixel(), both using an imageVector for the source feature. My ol.interaction.Select() uses the same idea as this example: [http://openlayers.org/en/latest/examples/select-features.html](url). Our attempt with the map.forEachFeatureAtPixel() is a very slightly modified versions of @ahocevar examples: [http://jsfiddle.net/ub2czho7/2/](url). Though we have also tried a modified version of this example of laying hover beneath text:[http://jsfiddle.net/jonataswalker/38e2hut1/1/](url) (which does not even perform well with a large number of features on Chrome on a Mac). 

I’m feeling like I’ve exhausted my resources on this, and maybe my only course of action is to add a catch to remove hover functionality for Windows machines. If needed I can post some custom fiddles, but I’ve not done much different from what’s been covered in this thread. I'm just hoping maybe someone has found a workaround recently.

**Comment by ahocevar** [comment]
If this is specific to Firefox on Windows: have you tried with a fresh profile without any plugin installed? If it is still slow, consider creating a bug report there. See the [Mozilla QA page](https://developer.mozilla.org/en-US/docs/Mozilla/QA) for more information.

**Comment by ahocevar** [comment]
Or try this modified version: http://jsfiddle.net/ub2czho7/7/.

One thing to investigate on the library side would be if hit detection is faster when done in an animation frame.

**Comment by BellisJ** [comment]
I'll check out your modified fiddle and see if I can make any improvements. But our performance seems to have randomly reached an acceptable level and begun working on Windows. At least, it has on a completely fresh Windows 10 install with a fresh Firefox install. Still need to try some more environments and see if we can replicate the crash again. I'll toss a bug report to Mozilla just to be safe.

**Comment by grwhumphries** [comment]
has this been resolved in any way, yet?  I'm having some major issues with my site being notably slower on anything but Chrome.  go to www.penguinmap.com/mapppd in firefox versus chrome to see what I'm talking about.  I'm sure it has something to do with the forEachFeatureatPixel() function.

**Comment by ahocevar** [comment]
A quick profiling session in Firefox shows that the application you mention suffers from other performance issues. 13% of the time is spent doing DOM manipulation. That comes from JQuery. 21% come from OpenLayers, but not from canvas rendering operations that cause the issue reported here. It's hard to tell what it is exactly because it uses a minified version of OpenLayers. But you can profile yourself using the Performance tab of the Firefox dev tools.

**Comment by grwhumphries** [comment]
![image](https://user-images.githubusercontent.com/1613298/29321598-db245b1c-81d2-11e7-8bcf-c82708f2ab56.png)

Can see now that Firefox is struggling primarily with DOM manipulation.  Why would that be? Works well in other browsers.

**Comment by ahocevar** [comment]
I would guess it's the tooltips/popups.

**Comment by fredj** [comment]
And also the `cursor` change. Try to change it only when it's needed

**Comment by grwhumphries** [comment]
Here's my code RE: tooltips and selecting points on click: 
`$(map.getViewport()).on('click', function (e) {

    var Drawing = false;
    map.getInteractions().forEach(function (interaction) {
        if (interaction instanceof ol.interaction.Draw) {
            Drawing = true;
            $('.PointSelectWarning').css('display', 'inline-block');
        } 
    });

    if (features.getLength() == 0 && Drawing == false) {        
        features.clear()
        selectedFeatures.clear()
        var pixel = map.getEventPixel(e.originalEvent);
        var feature = map.forEachFeatureAtPixel(pixel,
            function (feature, layer) {
                return feature;
            });
        if (feature) {           
            Fid = feature.getId()
            var Point = sourceGeoJson.getFeatureById(Fid);
            selectedFeatures.push(Point)

            URL = '/search/?p='.concat(Fid)

            getSearch(URL)
        }
    }
        
});`

**Comment by grwhumphries** [comment]
Yes - as you suspected.. when I turned off the above code, it worked fine

**Comment by Distem** [comment]
Hello everyone, good news I guess, like for most people in this thread firefox was mostly unusable with a large number of features screen, web workers for clustering by properties and complex styling - fps was averaging around **5**fps when panning (with all suggestions from this thread taking into account to improve the speed). Testing the latest **FF 55.0.3 (32-bit)** just randomly panning and zooming actually averaging **33**fps now (45 with 1 simple GeoJson LineString non-clustered layer).
![image](https://user-images.githubusercontent.com/7221884/29747722-39248b2a-8afb-11e7-9fb1-f169088376f0.png)
